components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +89,298 @@ public AbstractHTTPFunction() {
+                    if (result.getLeft() != HTTP_STATUS_TOO_MANY_REQUESTS &&
+                            result.getLeft() != HTTP_STATUS_UNAUTHORIZED &&
+                            !result.getRight().getLeft().equals(Constants.OUTCOME_SUCCESS)) {
+                        LOG.error("Error while calling endpoint. Url: " + endpointURL);
+                        result = executeRequestWithRetries(request, endpointURL, maxRequestAttemptsForAPIEndpointTimeout);
+                    }
+                    outcome = result.getRight().getLeft();
+                    JSONObject json = result.getRight().getRight();
+                    asyncReturn.accept(context, json != null ? json : Collections.emptyMap(), outcome);
+                }
+            } catch (Exception e) {
+                LOG.error("Error while applying authentication to the request.", e);
+                asyncReturn.accept(context, Collections.emptyMap(), Constants.OUTCOME_FAIL);
             }
+        });
+        JsGraphBuilder.addLongWaitProcess(asyncProcess, eventHandlers);
+    }
 
-            if (!isValidRequestDomain(request.getURI())) {
-                outcome = Constants.OUTCOME_FAIL;
-                LOG.error("Provided Url does not contain a allowed domain. Invalid Url: " + endpointURL);
-                asyncReturn.accept(context, Collections.emptyMap(), outcome);
-                return;
+    /**
+     * Execute the request with retries.
+     *
+     * @param request     HttpUriRequest.
+     * @param endpointURL Endpoint URL.
+     * @param maxRetries  Maximum number of retries.
+     * @return Pair of outcome and json.
+     */
+    private Pair<Integer, Pair<String, JSONObject>> executeRequestWithRetries
+    (HttpUriRequest request, String endpointURL, int maxRetries) {
+
+        Pair<Integer, Pair<String, JSONObject>> result = Pair.of(0, Pair.of(Constants.OUTCOME_FAIL, null));
+        JSONObject json = null;
+        String outcome = Constants.OUTCOME_FAIL;
+        int attempts = 0;
+
+        while (attempts < maxRetries) {
+            attempts++;
+            LOG.warn("Retrying the request. Attempt: " + attempts + " for endpoint: " + endpointURL);
+            if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                        DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                        Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.EXTERNAL_API, endpointURL)
+                        .resultMessage("Retrying the request for external endpoint.")
+                        .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                        .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
```

Comment:
Diagnostic log should also log the attempt

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/778ee19b0b9677f6d2cf96e829414272e804de01